### PR TITLE
Interior vehicle data rate limitation

### PR DESCRIPTION
**Pull request by ZhdanovP**
This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Provide Unit tests

### Summary
Have added limitation of requests for Interior vehicle data.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/get_interior_vehicle_data_request.cc [hunk 1]
```diff
@@ -117,6 +119,15 @@ void GetInteriorVehicleDataRequest::ProcessResponseToMobileFromCache(
   }
 }
 
+bool GetInteriorVehicleDataRequest::ProcessRateLimits() {
+  if (rcrpc_plugin_.amount_of_request_in_this_time_frame_ <
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/get_interior_vehicle_data_request.cc]: @ZhdanovP add traces with information about current amount of requests

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/get_interior_vehicle_data_request.cc [hunk 2]
```diff
@@ -147,6 +158,12 @@ void GetInteriorVehicleDataRequest::Execute() {
     return;
   }
 
+  if (!ProcessRateLimits()) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/get_interior_vehicle_data_request.cc]: @ZhdanovP if this function does not send response in the body please rename it to `CheckRateLimits`

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc [hunk 3]
```diff
@@ -53,11 +56,37 @@ bool RCRPCPlugin::Init(
...
+
   interior_data_cache_.reset(new InteriorDataCacheImpl());
+
+  const auto& settings = app_manager.get_settings();
+  const auto frequency = settings.get_interior_vehicle_data_frequency();
+  max_requests_count = frequency.first;
+
+  exit_timer_signal_ = std::promise<void>();
+
+  timer_thread = std::thread(
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc]: @ZhdanovP pease use for timer existing class Timer. Keep consistency in using utils with the project. 
The example you can find in Resumption : 

```
save_persistent_data_timer_("RsmCtrlPercist",
                                  new timer::TimerTaskImpl<ResumeCtrlImpl>(
                                      this, &ResumeCtrlImpl::SaveDataOnTimer))

...

void ResumeCtrlImpl::SaveDataOnTimer() {

```

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc [hunk 4]
```diff
@@ -53,11 +56,37 @@ bool RCRPCPlugin::Init(
...
+      frequency.second);
   return true;
 }
 
+RCRPCPlugin::~RCRPCPlugin() {
+  exit_timer_signal_.set_value();
+  timer_thread.join();
+}
+
+void RCRPCPlugin::TimerForResetRequestCount(const uint32_t time_frame) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc]: @ZhdanovP if this function will be timer callback, please rename it

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc [hunk 5]
```diff
@@ -53,11 +56,37 @@ bool RCRPCPlugin::Init(
...
+  std::future<void> futureObj = exit_timer_signal_.get_future();
+  while (futureObj.wait_for(std::chrono::milliseconds(1)) ==
+         std::future_status::timeout) {
+    std::this_thread::sleep_for(std::chrono::seconds(time_frame));
+    amount_of_request_in_this_time_frame_ = 0;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc]: @ZhdanovP looks like this variable may be accessed from different threads. It should be synchronized and wrapped by getters/setters, or getter and increment method.

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/test/commands/button_press_request_test.cc [hunk 6]
```diff
@@ -168,6 +169,7 @@ class ButtonPressRequestTest
       mock_allocation_manager_;
   testing::NiceMock<rc_rpc_plugin_test::MockInteriorDataCache>
       mock_interior_data_cache_;
+  rc_rpc_plugin::RCRPCPlugin rcrpc_plugin_;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/test/commands/button_press_request_test.cc]: Please do not use real classes ( dependencies) for testing of other classes. Ise mocks

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/test/commands/get_interior_vehicle_data_request_test.cc [hunk 7]
```diff
@@ -351,7 +353,7 @@ TEST_F(
           mobile_message);
 
   ON_CALL(mock_policy_handler_, CheckModule(_, _)).WillByDefault(Return(true));
-
+
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/test/commands/get_interior_vehicle_data_request_test.cc]: @ZhdanovP looks like style problem. Please check style

File: src/components/include/application_manager/rpc_plugins/rc_rpc_plugin/rc_settings.h [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#ifndef SRC_COMPONENTS_APPLICATION_MANAGER_RPC_PLUGINS_RC_RPC_PLUGIN_INCLUDE_RC_RPC_PLUGIN_RC_SETTINGS_H_
+#define SRC_COMPONENTS_APPLICATION_MANAGER_RPC_PLUGINS_RC_RPC_PLUGIN_INCLUDE_RC_RPC_PLUGIN_RC_SETTINGS_H_
+#include <utility>
+#include <stdint.h>
+
+namespace rc_rpc_plugin {
```
Comment (LuxoftAKutsan) [inline src/components/include/application_manager/rpc_plugins/rc_rpc_plugin/rc_settings.h]: @ZhdanovP this file redundant

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc (hunk 5 above)
Comment (ZhdanovP) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc]: @LuxoftAKutsan I have created this var like 'std::atomic<uint32_t>' for synchronization, the get/incr will be added.

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/commands/hmi/rc_button_press_response.h [hunk 9]
```diff
@@ -36,6 +36,7 @@
 #include "application_manager/commands/response_from_hmi.h"
 #include "rc_rpc_plugin/resource_allocation_manager.h"
 #include "rc_rpc_plugin/interior_data_cache.h"
+#include "rc_rpc_plugin/rc_rpc_plugin.h"
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/commands/hmi/rc_button_press_response.h]: @ZhdanovP why it was added?

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/commands/mobile/get_interior_vehicle_data_request.h.orig [hunk 10]
```diff
@@ -0,0 +1,111 @@
+/*
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/commands/mobile/get_interior_vehicle_data_request.h.orig]: @ZhdanovP redundant file

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/commands/mobile/get_interior_vehicle_data_request_BACKUP_18895.h [hunk 11]
```diff
@@ -0,0 +1,111 @@
+/*
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/commands/mobile/get_interior_vehicle_data_request_BACKUP_18895.h]: @ZhdanovP redundant file

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/interior_data_cache_impl.h [hunk 12]
```diff
@@ -34,23 +34,41 @@
...
   void Add(const std::string& module_type,
            const smart_objects::SmartObject& module_data) OVERRIDE;
   smart_objects::SmartObject Retrieve(
       const std::string& module_type) const OVERRIDE;
   bool Contains(const std::string& module_type) const OVERRIDE;
   void ClearCache() OVERRIDE;
+  void ResetRequestCountOnTimer() OVERRIDE;
+  uint32_t GetCurrentAmountOfRequests() const OVERRIDE;
+  void IncrementAmountOfRequests() OVERRIDE;
+  uint32_t GetMaxRequestCount() const OVERRIDE;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/interior_data_cache_impl.h]: @ZhdanovP should it be public?

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/interior_data_cache_impl.h [hunk 13]
```diff
@@ -34,23 +34,41 @@
...
+  InteriorDataCacheImpl(const uint32_t max_request_count,
+                        const uint32_t time_frame_of_allowed_requests,
+                        const uint32_t default_timeout);
   void Add(const std::string& module_type,
            const smart_objects::SmartObject& module_data) OVERRIDE;
   smart_objects::SmartObject Retrieve(
       const std::string& module_type) const OVERRIDE;
   bool Contains(const std::string& module_type) const OVERRIDE;
   void ClearCache() OVERRIDE;
+  void ResetRequestCountOnTimer() OVERRIDE;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/interior_data_cache_impl.h]: @ZhdanovP should it be public?

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/get_interior_vehicle_data_request.cc [hunk 14]
```diff
@@ -117,6 +118,20 @@ void GetInteriorVehicleDataRequest::ProcessResponseToMobileFromCache(
...
 }
 
+bool GetInteriorVehicleDataRequest::CheckRateLimits() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const uint32_t current_requests_amount =
+      interior_data_cache_.GetCurrentAmountOfRequests();
+  LOG4CXX_DEBUG(logger_,
+                "Current amount of requests in the same time frame is: "
+                    << current_requests_amount);
+  if (current_requests_amount < interior_data_cache_.GetMaxRequestCount()) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/get_interior_vehicle_data_request.cc]: @ZhdanovP max request count GetInteriorData can take from ini file, not from cache. It is not clear that cach contains max amount of requests just fo giving it to GetINteriorVD

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/interior_data_cache_impl.cc [hunk 15]
```diff
@@ -69,4 +82,27 @@ void InteriorDataCacheImpl::ClearCache() {
   sync_primitives::AutoLock autolock(subscriptions_lock_);
   subscriptions_.clear();
 }
+
+void InteriorDataCacheImpl::ResetRequestCountOnTimer() {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  bool is_alive = true;
+  while (is_alive) {
+    std::this_thread::sleep_for(
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/interior_data_cache_impl.cc]: @ZhdanovP timer is not a thread . I mean that do not implement cycle in timer. Timer implements cycle for you. Just reset amount of requests in timer

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc [hunk 16]
```diff
@@ -45,7 +45,13 @@ bool RCRPCPlugin::Init(
     application_manager::rpc_service::RPCService& rpc_service,
     application_manager::HMICapabilities& hmi_capabilities,
     policy::PolicyHandlerInterface& policy_handler) {
-  interior_data_cache_.reset(new InteriorDataCacheImpl());
+  const auto& settings = app_manager.get_settings();
+  const auto frequency = settings.get_vehicle_data_frequency();
+  interior_data_cache_.reset(
+      new InteriorDataCacheImpl(frequency.first,
+                                frequency.second,
+                                app_manager.get_settings().default_timeout()));
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_rpc_plugin.cc]: @ZhdanovP what is default timeout?

**Pull request merged**